src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs

```diff
@@ -66,6 +66,86 @@ public static void BuildChain()
+            yield return new object[] { false, X509ChainStatusFlags.UntrustedRoot | X509ChainStatusFlags.NotSignatureValid, BuildChainWithNotSignatureValidTest.UntrustedRoot };
+            yield return new object[] { false, X509ChainStatusFlags.PartialChain, BuildChainWithNotSignatureValidTest.PartialChain };
+            yield return new object[] { false, X509ChainStatusFlags.NotSignatureValid, BuildChainWithNotSignatureValidTest.EndCertNotSignatureValid };
+        }
+
+        [Theory]
+        [MemberData(nameof(BuildChainWithNotSignatureValidData))]
+        [PlatformSpecific(TestPlatforms.Linux)] // NotSignatureValid gets ignored on the root certificate.
+        public static void BuildChainWithNotSignatureValid(
+            bool chainBuildsSuccessfully,
+            X509ChainStatusFlags chainFlags,
+            BuildChainWithNotSignatureValidTest test)
+        {
+            TestDataGenerator.MakeTestChain3(
+                out X509Certificate2 endCert,
+                out X509Certificate2 intermediateCert,
+                out X509Certificate2 rootCert,
+                testName: test.ToString());
+
+            X509Certificate2 tampered;
+            switch (test)
+            {
+                case BuildChainWithNotSignatureValidTest.EndCertNotSignatureValid:
+                    tampered = TestDataGenerator.TamperSignature(endCert);
+                    endCert.Dispose();
+                    endCert = tampered;
+                    break;
+                default:
+                    // Make root cert signature invalid.
+                    tampered = TestDataGenerator.TamperSignature(rootCert);
+                    rootCert.Dispose();
+                    rootCert = tampered;
+                    break;
+            }
+
+            using (endCert)
+            using (intermediateCert)
+            using (rootCert)
+            using (ChainHolder chainHolder = new ChainHolder())
+            {
+                X509Chain chainTest = chainHolder.Chain;
+                chainTest.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                chainTest.ChainPolicy.VerificationTime = endCert.NotBefore.AddSeconds(1);
+                chainTest.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+                chainTest.ChainPolicy.ExtraStore.Add(endCert);
+
+                switch (test)
+                {
+                    case BuildChainWithNotSignatureValidTest.TrustedRoot:
+                    case BuildChainWithNotSignatureValidTest.EndCertNotSignatureValid:
+                        chainTest.ChainPolicy.ExtraStore.Add(intermediateCert);
+                        chainTest.ChainPolicy.CustomTrustStore.Add(rootCert);
+                        break;
+                    case BuildChainWithNotSignatureValidTest.PartialChain:
+                        chainTest.ChainPolicy.CustomTrustStore.Add(rootCert);
+                        break;
```

Comment:
* "I think we want a test where PartialChain and NotSignatureValid come out at the same time"
* "Uh, OK.  But, um, if we're PartialChain, how did we know what the public key was that couldn't verify the signature?"
* "Oh.  Good point."

There's a possibility that a "self-issued, but not self-signed" certificate might hit that case... but while one or both of X.509 and RFC5280 call that a CA key rollover scenario I don't think it's generally well handled (e.g. OpenSSL by default skips the distinction between self-issued and self-signed).

Reply:
Actually, we could get it, but from a different shape of test:

* Make chain 3
* Tamper endEntity
* Don't add rootCert to either the trust store or ExtraStore

That shows that the overall chain is capable of producing both values.